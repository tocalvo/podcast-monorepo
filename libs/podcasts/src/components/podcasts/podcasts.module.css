.container {
  display: flex;
  flex-direction: column;
  margin: 10px;
}

.form {
  display: flex;
  flex-direction: row;
  justify-content: flex-end;
}
.number {
  background-color: #4682b4;
  padding: 0px 8px;
  color: white;
  border-radius: 2%;
}
.filter {
  margin-left: 10px;
  border: 1px solid grey;
}
.podcasts-grid {
  display: flex;
  flex-wrap: wrap;
  gap: 10px;
  justify-content: space-evenly;
  margin-top: 50px;
}
