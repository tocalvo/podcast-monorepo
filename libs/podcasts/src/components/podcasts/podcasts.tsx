import { useItunesPodcasts } from '../../hooks';
import styles from './podcasts.module.css';
import PodcastCard from '../podcastCard/podcastCard';
import { useMemo, useState } from 'react';

export const Podcasts: React.FC = () => {
  const podcastsQuery = useItunesPodcasts();

  const entries = useMemo(
    () => podcastsQuery.data?.feed.entry || [],
    [podcastsQuery.data]
  );
  const [filter, setFilter] = useState('');

  const filteredEntries = useMemo(() => {
    const lowerCaseFilter = filter.toLowerCase();
    return entries.filter(
      (entry) =>
        entry['im:name'].label.toLowerCase().includes(lowerCaseFilter) ||
        entry['im:artist'].label.toLowerCase().includes(lowerCaseFilter)
    );
  }, [entries, filter]);

  return (
    <div className={styles['container']}>
      <div className={styles['form']}>
        <p data-testid="num-podcasts" className={styles['number']}>{filteredEntries.length}</p>
        <input
          className={styles['filter']}
          type="text"
          placeholder="Filter podcasts..."
          value={filter}
          onChange={(e) => setFilter(e.target.value)}
        ></input>
      </div>
      <div className={styles['podcasts-grid']}>
        {filteredEntries.map((entry, i) => (
          <PodcastCard key={i} podcast={entry} />
        ))}
      </div>
    </div>
  );
};

export default Podcasts;
