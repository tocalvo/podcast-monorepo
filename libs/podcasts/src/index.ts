export * from './components';
export * from './hooks';
export * from './__mocks__';
export * from './models';
