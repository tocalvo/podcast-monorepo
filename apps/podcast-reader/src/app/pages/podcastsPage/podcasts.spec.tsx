import { fireEvent, render, screen } from '@testing-library/react';

import Podcasts from './podcastsPage';
import { BrowserRouter } from 'react-router';
import { podcastListMock } from '@org/podcasts';

vi.mock('@org/podcasts', async () => ({
  ...(await vi.importActual('@org/podcasts')),
  useItunesPodcasts: vi.fn(() => ({
    podcasts: podcastListMock.feed.entry,
    getPodcastResumeById: vi
      .fn()
      .mockReturnValue(podcastListMock.feed.entry[0]),
  })),
}));

describe('Podcasts', () => {
  it('should render successfully', () => {
    const { baseElement, container } = render(
      <BrowserRouter>
        <Podcasts />
      </BrowserRouter>
    );
    expect(baseElement).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  it('should render one card by podcast', () => {
    render(
      <BrowserRouter>
        <Podcasts />
      </BrowserRouter>
    );
    const cards = screen.getAllByTestId('podcast-card'); // Asegúrate de agregar el data-testid="podcast-card" en PodcastCard
    expect(cards).toHaveLength(2);
  });

  it('should filter the cards using their input', () => {
    render(
      <BrowserRouter>
        <Podcasts />
      </BrowserRouter>
    );
    const initCards = screen.getAllByTestId('podcast-card'); // Asegúrate de agregar el data-testid="podcast-card" en PodcastCard
    expect(initCards).toHaveLength(2);
    fireEvent.change(screen.getByPlaceholderText('Filter podcasts...'), {
      target: { value: 'Joe Budden' },
    });
    const filteredCards = screen.getAllByTestId('podcast-card');
    expect(filteredCards).toHaveLength(1);

    expect(screen.queryByText('The Joe Budden Podcast')).toBeTruthy();
    expect(screen.queryByText('New Rory')).toBeNull();
  });

  it('should show the podcast number', () => {
    render(
      <BrowserRouter>
        <Podcasts />
      </BrowserRouter>
    );
    expect(screen.getByTestId('num-podcasts').textContent).toBe('2');
    fireEvent.change(screen.getByPlaceholderText('Filter podcasts...'), {
      target: { value: 'Joe Budden' },
    });
    expect(screen.getByTestId('num-podcasts').textContent).toBe('1');
  });
});
