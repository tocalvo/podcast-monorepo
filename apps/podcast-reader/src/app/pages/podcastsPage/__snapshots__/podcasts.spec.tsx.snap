// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Podcasts > should render successfully 1`] = `
<div>
  <div
    class="_container_10e0b0"
  >
    <div
      class="_form_8aea71"
    >
      <p
        class="_number_8aea71"
        data-testid="num-podcasts"
      >
        2
      </p>
      <input
        class="_filter_8aea71"
        placeholder="Filter podcasts..."
        type="text"
        value=""
      />
    </div>
    <div
      class="_podcasts-grid_10e0b0"
    >
      <a
        aria-label="Ir al detalle del podcast"
        class="_card_bbe757"
        data-discover="true"
        data-testid="podcast-card"
        href="/podcast/1535809341"
      >
        <img
          alt="podcast logo"
          class="_image_bbe757"
          src="https://is1-ssl.mzstatic.com/image/thumb/Podcasts113/v4/f2/21/fa/f221fabd-017f-5125-633b-f1fe4f39802a/mza_182995249085044287.jpg/55x55bb.png"
        />
        <p
          class="_title_bbe757"
        >
          The Joe Budden Podcast
        </p>
        <p
          class="_artist_bbe757"
        >
          Author: 
          The Joe Budden Network
        </p>
      </a>
      <a
        aria-label="Ir al detalle del podcast"
        class="_card_bbe757"
        data-discover="true"
        data-testid="podcast-card"
        href="/podcast/1572182022"
      >
        <img
          alt="podcast logo"
          class="_image_bbe757"
          src="https://is1-ssl.mzstatic.com/image/thumb/Podcasts112/v4/31/80/56/3180562e-ac0b-f10e-7120-641b1c26e0f8/mza_10135383815267163613.jpg/55x55bb.png"
        />
        <p
          class="_title_bbe757"
        >
          New Rory & MAL
        </p>
        <p
          class="_artist_bbe757"
        >
          Author: 
          Rory Farrell & Jamil "Mal" Clay & Studio71
        </p>
      </a>
    </div>
  </div>
</div>
`;
